src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsBuildLogParseUtils.java

```diff
@@ -0,0 +1,81 @@
+import org.jsoup.nodes.Node;
+import org.jsoup.nodes.TextNode;
+
+import de.tum.in.www1.artemis.domain.BuildLogEntry;
+
+public class JenkinsBuildLogParseUtils {
+
+    // Pattern of the DateTime that is included in the logs received from Jenkins
+    private static final DateTimeFormatter LOG_DATA_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssX");
+
+    public static List<BuildLogEntry> parseBuildLogsFromJenkinsLogs(List<String> logLines) {
+        final List<BuildLogEntry> buildLogs = new ArrayList<>();
+        for (final var logLine : logLines) {
+            // The build logs that we are interested in are the ones that start with a timestamp
+            // of format [timestamp] ...
+            final String possibleTimestamp = StringUtils.substringBetween(logLine, "[", "]");
+            if (possibleTimestamp == null) {
+                continue;
+            }
+
+            try {
+                final ZonedDateTime timestamp = ZonedDateTime.parse(possibleTimestamp);
+                final String log = logLine.substring(possibleTimestamp.length() + 2);
+
+                BuildLogEntry buildLogEntry = new BuildLogEntry(timestamp, stripLogEndOfLine(log).trim());
+                buildLogs.add(buildLogEntry);
+            }
+            catch (DateTimeParseException e) {
+                // The log line doesn't contain the timestamp so we ignore it
+            }
+        }
+        return buildLogs;
+    }
+
+    public static List<BuildLogEntry> parseLogsLegacy(Element logHtml) {
+        final var buildLog = new LinkedList<BuildLogEntry>();
+        final var iterator = logHtml.childNodes().iterator();
+        while (iterator.hasNext()) {
+            final var node = iterator.next();
+            final String log;
+            // For timestamps, parse the <b> tag containing the time as hh:mm:ss
+            if (node.attributes().get("class").contains("timestamp")) {
+                final var timeAsString = ((TextNode) node.childNode(0).childNode(0)).getWholeText();
+                final var time = ZonedDateTime.parse(timeAsString, LOG_DATA_TIME_FORMATTER);
+                log = reduceToText(iterator.next());
+                buildLog.add(new BuildLogEntry(time, stripLogEndOfLine(log)));
+            }
+            else {
+                // Log is from the same line as the last
+                // Look for next text node in children
+                log = reduceToText(node);
+                final var lastLog = buildLog.getLast();
+                lastLog.setLog(lastLog.getLog() + stripLogEndOfLine(log));
+            }
+        }
+        return buildLog;
+    }
+
+    private static String stripLogEndOfLine(String log) {
+        return log.replaceAll("[\\r\\n]", "");
+    }
+
+    private static String reduceToText(Node node) {
+        if (node instanceof TextNode) {
+            return ((TextNode) node).getWholeText();
+        }
+
+        return reduceToText(node.childNode(node.childNodeSize() - 1));
```

Comment:
Maybe add a comment, why you use the last child node here.

Reply:
I honestly don't know what this code is doing.It's used in the `parseLogsLegacy` function which is used for parsing legacy build logs in HTML format. 

I introduced a new function in this PR called `parseBuildLogsFromJenkinsLogs` which parses logs from strings instead of HTML. I use that for parsing pipeline logs and would like to use it for parsing legacy logs too but I sadly I can't test the changes because I don't know what "legacy" means.